### ISSUE TITLE
Selecting Multiple Provisioners

### ISSUE THREAD

**Issue by ellistarn** [awslabs/karpenter#783]
> Note: This issue is extracted from @stevehipwell's comment on https://github.com/awslabs/karpenter/issues/628

@ellistarn sorry for the long response, I've been thinking about this since I first saw Karpenter and compared it to our current solution (see below). The following spec might seem like a big change from your current design but I think it just shifts the configuration into multiple provisioners.

Primary constraints:

Provisioners are given a priority weight
Provisioners declare labels, either directly or based on options such as instance type
Provisioners declare taints
Provisioners can add additional taints to a node via a custom Karpenter pod annotation
Provisioner selection logic:

Provisioners are ordered by weight
First provisioner to have required labels from pod spec and have tolerated taints is selected
All equally weighted provisioners are evaluated and if suitable added to the first found one
Potential provisioners are evaluated to pick a winning spec
Traits:

Can use a single provisioner for simple cases
Only need extra provisioners for pooling nodes by type
Portable pod specifications
Can deploy same application in either ASG & CA or Karpenter cluster
Tolerations only have impact if desired
Fully customisable for complex workloads
No need to tightly couple spec to specific provisioner with custom annotation
An example use-case would be local SSD storage. Given 3 provisioners:

- name: gp
  weight: 50
  instances: [m5]
  labels: []
  taints: []
- name: gp-ssd
  weight: 50
  instances: [m5d]
  labels: [ssd=true]
  taints: []
- name: storage-optimised
  weight: 100
  instances: [i3]
  labels: [storage-optimised=true, ssd=true]
  taints: [storage-optimised]
If a general pod is scheduled it might get a m5 or a m5d instance. Then if a pod with a ssd=true label selector is scheduled it could go onto an existing gp-ssd provisioned node or provision a new one. If the pod tolerates storage-optimised it could also be scheduled onto a node from the storage-optimised provisioner. If the node had a storage-optimised=true selector it could only go onto a node from the storage-optimised provisioner.

There needs to be thought given to affinity/anti-affinity rules as this probably need to be retrospective to the scheduler unless you can use a webhook to intercept them before a pod is scheduled. I'm specifically thinking about a multi region soft anti-affinity.

It'd also be great if Karpenter could also add custom labels such as a subnet identifier (there are no default labels suitable for custom networking if you have public and private subnets), presence of SSDs (wouldn't need a custom provisioner for m5d types to select them based on this).

For reference our current solution is built as groups of 6 ASGs (an ASG per AZ duplicated between on-demand & spot) created per node pattern, which is based on type (e.g. gp, compute, network, etc), size (large, xlarge, etc) and labels/taints. These ASGs are scaled by Cluster Autoscaler using the priority expander. This pattern get us to about 80% of what we'd like, Karpenter working as I've described above would hit close to 100%.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by ellistarn** [comment]
I think there are multiple interesting interrelated ideas being discussed here. I tried to understand your request and translate it to use cases we can work backwards from to identify a solution (potentially your recommendation).

Mechanisms under discussion:

1. Taint Generation
2. Provisioner Selection
3. Label Generation

### Taint Generation
Karpenter currently applies taints to launched nodes from both `provisioner.spec.taints` and generated for `pod.spec.tolerations`. According to the taints semantic, a pod with tolerations is schedulable whether or not the taints are applied any given node. However, in the current iteration of Karpenter, we chose to conservatively generate taints for pod tolerations in all scenarios.

**Missing Use Case:** I can define a pod that can schedule to existing tainted nodes, but is not otherwise isolated. 

The example you provided is `CriticalAddonsOnly` which is applied to many EKS addons. Pods containing this toleration should schedule to existing capacity, but if they can't fit, can be binpacked with other noncritical pods. I *think* what you're asking for is the opportunity for tighter binpacking for these pods. Can you elaborate more on this use case? 

The main reason we choose to generate taints is to reduce the number of provisioners that users have to define (in case additional isolation is desired) and we didn't have a logical example of the missing use case you describe. The main reason for wanting to reduce the number of provisioners is because of the bespoke nature of the selection logic (as you point out), which brings us to...

### Provisioner Selection
Karpenter currently has a concept of a `default` provisioner, which is used for all pods. Pods can specify a node selector (e.g. `karpenter.sh/provisioner-name=another-provisioner`) to select a different provisioner. Rules are merged from the selected provisioner and the pod spec, allowing provisioners to specify a set of baseline scheduling rules so they don't need to be repeated in every pod spec. 

In current NodeGroup + CAS setups, you might put a label like `has-fast-storage=true` on your node group and then target that label using a nodeSelector on a given pod. You can layer abstract concepts (like fast storage) on multiple node groups with different shapes and properties. 

In karpenter's current design, you're must either express these requirements explicitly at the pod level (which can be verbose/repetitive), or at the provisioner level and use a bespoke `karpenter.sh/provisioner-name` selector to get those rules. This requires pods to be aware of infrastructure level configuration, which can be difficult to coordinate when provisioners change.

**Missing Use Case:** I can define abstract node requirements and select them using labels that I define.

This is definitely a gap in the current design. This is currently complicated by the fact that if you were to attempt to define labels like `has-fast-storage=true`, karpenter would happily generate those labels for you, but without any of the semantic meaning you're attempting to ascribe to it. This brings us to...

### Label Generation
Karpenter currently applies labels to launched nodes using both `provisioner.spec.labels`, generated from `pod.spec.nodeSelector/nodeAffinity`. The idea behind custom label generation is again to limit the number of provisioners that must be defined and allow abstract concepts (unrelated to actual node attributes) to be added to nodes from the pod level. 

For example, I can add a `team=frontend` nodeSelector to all of my pods and know that the nodes they run on will be identifiable for my team. However, as I think more about this, that there may be some strange effects where if those pods finish and other pods are scheduled there, there may be some drift. 

**Confusing Behavior**: Custom labels on nodes may become nonsensical over the lifetime of the node.

It's core to karpenter's design that pods can specify "well known labels" e.g. `topology.kubernetes.io/zone` to express concrete runtime constraints, but perhaps "custom labels", which express abstract concepts, should instead be defined the provisioner level.

### Additional thoughts
- We have played with the idea of Karpenter becoming a custom scheduler and using the `pod.spec.schedulerName` field. There are some advantages to this like avoiding race conditions with the kube-scheduler. It hasn't been implemented so far due to the effort required. The approach you mention would prevent us from taking this road in the future. 
- I've thought about selecting provisioners using namespaces. Something like `provisioner.spec.podNamespaces` or perhaps even `namespace.metadata.labels[karpenter.sh/provisioner-name]=default`. 
- We currently support preferential fallback at the pod level (i.e. pod.spec.affinity.nodeAffinity.preferred...) and prioritized provisioners give us an interesting path towards provisioner-level preferential fallback. 
- What you're proposing brings Provisioners to be much more similar to flexible NodeGroups, which the selection logic playing a similar role to CAS. I think there may be some advantages to this. 

I hope that this sounds in the same ballpark to what you're thinking. I need to spend some more time thinking about the side effects of pods applying to multiple provisioners. I've thought about similar things in the future and had concluded it was a dead end -- but I can't remember why.

**Comment by ellistarn** [comment]
FYI @stevehipwell I'm starting to prototype this discussion: https://github.com/awslabs/karpenter/pull/791

**Comment by stevehipwell** [comment]
Thanks for the effort and work here @ellistarn and apologies for taking my time to respond, I wanted to give this my full attention. I think you've summed up my points very succinctly, but I'll try and add a few more use cases to expand on why the behaviour is desirable and at least in my case required.

---

### Taint generation

The fundamental point here is that pod tolerations shouldn't have any bearing on the node spec as they are permissive rather than declarative. If we exclude Karpenter from the picture a toleration has no impact on a system without taints; so Karpenter should follow this high level rule to be consistent.

That said Karpenter gives us the ability to design a system to send workloads to tainted nodes, but this is more complex that it seems at first glance. For this to work correctly (deterministically) it needs to both taint & label a node and make sure the pod has the correct selector and tolerations. This could be done by validating a custom Karpenter annotation references an defined toleration and selector but it seems much more suited to a webhook to modify the pod spec based on the annotation.

#### Use cases

- Tolerations only matter when I'm using taints and I will manage them at that point
- I don't want to have to manage tolerations coming from Helm charts if I'm not using taints

#### Stretch use cases

- I want to be able to isolate a pod (on its own or with other pods configured the same way) without requiring a separate provider

### Label generation

I've moved label generation above provisioner selection as it can be used to reduce the number of provisioners required. Provisioners already add custom well known labels such as zone etc but this would be relatively easy to extend to cover additional use cases (e.g. local SSDs, subnet id, etc). This could be further extended to be user defined, but the same could be achieved by adding provisioners which is probably simpler.

Following this model the only labels on a node would be driven by the provisioner so if they're outdated it's because the provisioner has been removed but the node hasn't.

#### Use cases

- I'd like my node to have labels for additional capabilities (e.g. local ssd)
- I'd like my node to have labels describing it's context (e.g. subnet id)

### Provision selection

I agree that my suggestions operate more like a flexible node group, although I think 1 provisioner could replace at least 6 node groups in my use case. However I'd prefer to pitch it as an abstraction allowing the provisioners to be provided/managed/maintained separately from the workloads running on top of the provisioned nodes. This open/closed design fits nicely with the way the Kubernetes architecture is layered and would support PaaS use of this.

#### Use cases

- I want to be able to manage provisioners for a cluster independently from the workloads
- I want workloads to be able to target know labels and taints I've documented without knowing how it's implemented
- I want to select subnets by provisioner (this might already be present?)

### Additional use cases

- I'd like provisioners to support soft multi zone affinity without preemptable pods
- I'd like to define a node age lifecycle via a provisioner
- I'd like nodes to be removed (opt out maybe?) when the provisioner that created them is removed

**Comment by ellistarn** [comment]
Hey @stevehipwell, this just got merged! https://github.com/awslabs/karpenter/pull/803

**Comment by stevehipwell** [comment]
@ellistarn are there any docs to go with these changes?

**Comment by ellistarn** [comment]
We will be writing up some documentation and cutting a release next week. 

The tl;dr is that we currently select randomly if multiple provisioners are matched. I know this isn't yet the full vision of what we discussed, but it's an incremental step from the previous design. Provisioners may be selected via user defined labels, rather than `karpenter.sh/provisioner-name`. I currently recommend that you don't overlap labels defined on provisioner to maximize binpacking opportunities. 

The remaining chunk of scope I think of as "provisioner preferential fallback". This could be priority or some other system. I'm not I'd like to continue to think about priority or maybe something else. I'd love to have this conversation or hear any of your other thoughts in this space.
